src/API.jl

```diff
@@ -895,31 +895,60 @@ end
 precompile() = precompile(Context())
 function precompile(ctx::Context)
     printpkgstyle(ctx, :Precompiling, "project...")
-
-    pkgids = [Base.PkgId(uuid, name) for (name, uuid) in ctx.env.project.deps if !is_stdlib(uuid)]
+    
+    num_tasks = parse(Int, get(ENV, "JULIA_NUM_PRECOMPILE_TASKS", string(Sys.CPU_THREADS + 1)))
+    parallel_limiter = Base.Semaphore(num_tasks)
+    
+    man = Pkg.Types.read_manifest(ctx.env.manifest_file)
+    pkgids = [Base.PkgId(first(dep), last(dep).name) for dep in man if !Pkg.Operations.is_stdlib(first(dep))]
+    pkg_dep_uuid_lists = [collect(values(last(dep).deps)) for dep in man if !Pkg.Operations.is_stdlib(first(dep))]
+    filter!.(!is_stdlib, pkg_dep_uuid_lists)
+    
     if ctx.env.pkg !== nothing && isfile( joinpath( dirname(ctx.env.project_file), "src", ctx.env.pkg.name * ".jl") )
         push!(pkgids, Base.PkgId(ctx.env.pkg.uuid, ctx.env.pkg.name))
+        push!(pkg_dep_uuid_lists, collect(keys(ctx.env.project.deps)))
+    end    
+    
+    precomp_events = Dict{Base.UUID,Base.Event}()
+    was_recompiled = Dict{Base.UUID,Bool}()
+    for pkgid in pkgids
+        precomp_events[pkgid.uuid] = Base.Event()
+        was_recompiled[pkgid.uuid] = false
     end
-
+    
+    precomp_tasks = Task[]
     # TODO: since we are a complete list, but not topologically sorted, handling of recursion will be completely at random
-    for pkg in pkgids
+    for (i, pkg) in pairs(pkgids)
         paths = Base.find_all_in_cache_path(pkg)
         sourcepath = Base.locate_package(pkg)
         sourcepath === nothing && continue
         # Heuristic for when precompilation is disabled
         occursin(r"\b__precompile__\(\s*false\s*\)", read(sourcepath, String)) && continue
-        stale = true
-        for path_to_try in paths::Vector{String}
-            staledeps = Base.stale_cachefile(sourcepath, path_to_try)
-            staledeps === true && continue
-            # TODO: else, this returns a list of packages that may be loaded to make this valid (the topological list)
-            stale = false
-            break
-        end
-        if stale
-            Base.compilecache(pkg, sourcepath)
-        end
+        
+        t = @async begin
+            length(pkg_dep_uuid_lists[i]) > 0 && wait.(map(x->precomp_events[x], pkg_dep_uuid_lists[i]))
```

Comment:
`wait.(...)` creates unnecessary `Vector{Nothing}`. You can do `foreach(wait, ...)` but at this point it's probably better to jsut write a `for` loop:

```suggestion
            for x in pkg_dep_uuid_lists[i]
                wait(precomp_events[x])
            end
```